### DISCUSSION TITLE
Recommendations for where to override CRUD logic

### DISCUSSION THREAD

**Discussion by lovetoburnswhen** [aminalaee/sqladmin#69] [Q&A]
I was wondering if there were any good override points for introducing custom logic into the CRUD operations. 

Ideally there would be something like `ModelAdmin._custom_create(self, **kwargs)` that would receive the validated form data and would be separate from  any HTTP/user-facing stuff

**Comment by aminalaee** [comment]
Hey,

Just to be clear, why do you need to implement the whole create logic? Why won't some events like `ModelAdmin.after_model_created(self, model)` help you add your logic?

And at the moment we don't have either of those, I was just wondering what your use-case is.

**Reply by lovetoburnswhen** [reply]
My app funnels all CRUD operations through a set of `CRUDServices` which handle things like logging, audit trail/history table, cascades, etc. 

Ideally I'd be able to replace `MyModel(**kwargs)` with `MyModelService.create(**kwargs)` so CRUD operations performed via the admin GUI utilize the same code as those performed through the API

**Reply by aminalaee** [reply]
Yeah this makes sense now. I just need to figure out how to handle both sync and async versions, probably the easiest way is to have `async def create` and `def create_sync` which the SQLAdmin will call both but you can implement either based on your case.

**Reply by lovetoburnswhen** [reply]
Another way could be to break up the `Admin.create()` method which currently does a lot of stuff internally. Maybe even something as simple as `self._get_form_context -> {"request": ..., "model_admin": ..., "form": ...}`. The remainder is small enough to duplicate without too much overhead.
